### PR TITLE
Fix Issue 17899: Allow delegate initialisation at module scope

### PR DESCRIPTION
**Pull request by thewilsonator**
~~Delegates are being codegenned with `.ptr` and `.funcptr` swapped, but otherwise working as intended~~ Fixed, all green.

### CODE REVIEW

**Comment by thewilsonator** [comment]
I think I might change the test from a compilable to a runnable with some checkable side effect tomorrow.

File: src/dmd/todt.d
Comment (wilzbach) [inline src/dmd/todt.d]: Not covered.

Comment (thewilsonator) [reply src/dmd/todt.d]: True, I dont know what would trigger it though. I'll just remove it.

File: test/runnable/test17899.d
Comment (wilzbach) [inline test/runnable/test17899.d]: Test failed.  The logged output:
/home/braddr/sandbox/at-client/pull-2994750-FreeBSD_32/dmd/generated/freebsd/release/32/dmd -conf= -m32 -Irunnable   -odgenerated/runnable -ofgenerated/runnable/test17899_0 runnable/test17899.d
generated/runnable/test17899_0


==============================
Test failed: caught signal: 11

Comment (thewilsonator) [reply test/runnable/test17899.d]: adding 
```d
    writeln(cast(void*)bar1.funcptr," ",cast(void*)bar1.ptr);
    writeln(cast(void*)bar2.funcptr," ",cast(void*)bar2.ptr);
```
yields 
```
null 10BD59174
null 10BD591B4
```
Looks like the `.funcptr` is being initialised with the `.ptr` and visa versa.
Any help much appreciated.

Comment (thewilsonator) [reply test/runnable/test17899.d]: yep, they are the wrong way round.
```d
void swap(T)(ref T t)
{
    auto tmp = t.ptr;
    t.ptr = cast(void*)t.funcptr;
    t.funcptr = cast(typeof(t.funcptr)) tmp;
}
void main()
{
    assert(setme  == 0);
    swap(bar1);
    bar1();
    assert(setme == 1);
    swap(bar2);
    bar2();
    assert(setme == 2);
}
```
works.

Comment (thewilsonator) [reply test/runnable/test17899.d]: Added padding so that `funcptr` should align with the runtime offset. Hopefully this works.

**Comment by thewilsonator** [comment]
@wilzbach this is no longer WIP, its green and ready to merge unless there are any objections. Do I need to squash the commits?

**Comment by wilzbach** [comment]
Nope it's a missing feature of the dlang-bot to remove "WIP" automatically:

>  dlang-bot added the WIP label 2 days ago
> @thewilsonator changed the title from [WIP] Issue 17899: Allow delegate initialisation at module scope to Fix Issue 17899: Allow delegate initialisation at module scope 2 days ago

(I just removed it).

**Review by JinShil — DISMISSED** [review]
Seems like a good idea to me, but I would feel more comfortable having feedback from a more senior DMD team member.

**Comment by wilzbach** [comment]
(rebased to remove the failing Travis build from the CI list)

**Comment by thewilsonator** [comment]
Thanks!

File: test/fail_compilation/ice13259.d
Comment (wilzbach) [inline test/fail_compilation/ice13259.d]: Move the test to `compilable`?

Comment (thewilsonator) [reply test/fail_compilation/ice13259.d]: I should be superseded by the other test I added. But I can if need be.

Comment (wilzbach) [reply test/fail_compilation/ice13259.d]: Ehm then just copy paste this line into the other test. I think ICEs are special as we definitely don't want any regressions there.

Comment (thewilsonator) [reply test/fail_compilation/ice13259.d]: done.

**Comment by thewilsonator** [comment]
Yay, its green! Do I need to add a change log for this?

**Comment by ibuclaw** [comment]
Yes, you would.

**Comment by ibuclaw** [comment]
Not that I want to be pedantic, but do you have any _useful_ tests or rationale for this change?

**Comment by ibuclaw** [comment]
That is to say: Why aren't you using a function type? Delegates make little sense if declared at module level.

**Comment by thewilsonator** [comment]
My personal use is the callback hander for compute API errors in DCompute. The default action is to throw an error, but if a user is anticipating an error they may set the delegate to something appropriate for the situation.

But it is useful for any module level delegate that is desired to be initialised.

**Comment by thewilsonator** [comment]
Anyone know why the some of the test are now failing?

**Comment by JinShil** [comment]
> Anyone know why the some of the test are now failing?

The auto-tester failure appears to be related to #7747, but I don't know why as the tests all passed before that was merged.  It's being reverted in #7755, so probably best to wait until that is merged before checking again.

**Comment by thewilsonator** [comment]
I don't know what more justification I can give other than "I need this as a callback, functions are pretty crap at that": removes a useless limitation of the compiler and the cryptic error message that it produces? You can already declare delegates at module scope, you just can't initialise them (even with module constructors because of a bug that I have yet to diagnose properly, but in theory you can initialise the with a module constructor). Put it this, way imagine this were a pr to remove the ability to initialise delegates (completely ignoring the code breakage that may occur for the sake of argument) declared at module scope because "you [c|sh]ould be using a function for this".

I mean I suppose you could use a function that you set to call a function local delegate but that's just silly when you could set a delegate instead.

/rant

**Comment by andralex** [comment]
@thewilsonator thanks!! can you please change the language specification accordingly?

**Comment by thewilsonator** [comment]
Which section: 19.19, 10.21.10, both?

**Comment by andralex** [comment]
19.19 please

**Comment by thewilsonator** [comment]
done -> https://github.com/dlang/dlang.org/pull/2119

**Pull request merged**
